### PR TITLE
Preliminary PR [Don't merge]

### PR DESCRIPTION
**Pull request by shunshou**
@grebe Attempt at adding Interval type class. Would appreciate a quick look. :)
+ Chick is adding optimization hooks.

### CODE REVIEW

File: src/main/scala/dsptools/numbers/chisel_types/IntervalTypeClass.scala [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+  def shr(a: Interval, n: Int): Interval = a >> n
+  def shr(a: Interval, n: UInt): Interval = a >> n
+
+  // mul2 consistent with shl
+  // signBit relies on Signed
+
+  // Retains significant digits while dividing
+  override def div2(a: Interval, n: Int): Interval = {
+    require(a.binaryPoint.known, "div2 Binary point must be known!")
+    val outFull = a.shiftLeftBinaryPoint(n) >> n
```
Comment (grebe) [inline src/main/scala/dsptools/numbers/chisel_types/IntervalTypeClass.scala]: Maybe I'm just confused- is the `>>` not removing significant digits here?

File: src/main/scala/dsptools/numbers/chisel_types/UIntTypeClass.scala [hunk 2]
```diff
@@ -34,7 +34,7 @@ trait UIntRing extends Any with Ring[UInt] with hasContext {
     }
     ShiftRegister(diff.asUInt, context.numAddPipes)
   }
-  def negate(f: UInt): UInt = -f
+  def negate(f: UInt): UInt = throw DspException("Can't negate UInt and get UInt")
```
Comment (grebe) [inline src/main/scala/dsptools/numbers/chisel_types/UIntTypeClass.scala]: Should this be an exception? Minus is supported. Negation in the sense of additive inverse still makes sense for `UInt`s.

File: src/main/scala/dsptools/tester/VerilogTbDump.scala [hunk 3]
```diff
@@ -25,7 +25,8 @@ trait VerilogTbDump {
 
   val (inputs, outputs) = TestersCompatibility.getDataNames("io", dut.io) partition {
     case (dat, name) =>
-      DataMirror.directionOf(dat) == chisel3.core.ActualDirection.Input
+      dat.dir == core.Direction.Input
+      //DataMirror.directionOf(dat) == chisel3.core.ActualDirection.Input
```
Comment (grebe) [inline src/main/scala/dsptools/tester/VerilogTbDump.scala]: What's going on here- is `DataMirror` changing again?

File: src/main/scala/dsptools/numbers/chisel_types/IntervalTypeClass.scala (hunk 1 above)
Comment (shunshou) [reply src/main/scala/dsptools/numbers/chisel_types/IntervalTypeClass.scala]: Yeah uhh that's confusing -- just following Adam's firrtl convention. >> removes significant digits, so you first have to LSB pad  binary points with slbp first :\

File: src/main/scala/dsptools/numbers/chisel_types/UIntTypeClass.scala (hunk 2 above)
Comment (shunshou) [reply src/main/scala/dsptools/numbers/chisel_types/UIntTypeClass.scala]: Even though it's not representable as a UInt?

File: src/main/scala/dsptools/tester/VerilogTbDump.scala (hunk 3 above)
Comment (shunshou) [reply src/main/scala/dsptools/tester/VerilogTbDump.scala]: No it didn't change, but the version of Chisel with ranges doesn't have DataMirror stuff, so I had to revert back until ranges get merged into Chisel master :(...

**Comment by shunshou** [comment]
Turns out I have syntax errors in my pushes -- fixing!

**Comment by shunshou** [comment]
@chick @grebe -- when you guys wrote tests, were there tests that passed that were maybe incorrect? i.e. dsptools.numbers.FixedPrecisionChangerSpec has one test that keeps failing + another test that is right but should be failing...

**Comment by shunshou** [comment]
@chick can I request a favor? I'm trying to get matrix ops working with:
sbt "test-only dsptools.toys.MatrixOpSpec"

But the last test -- the one using DspReal fails, because the input doesn't seem to get passed to the output. It's really weird because the generated FIRRTL looks right and your instrumentation csv file shows that the poke value is getting to the input to my "TestModule"... but both interpreter and verilator (from the VCD files) seem to think that the TestModule input isn't getting propagated to the actual MatrixOp module... (even though the a.node stuff propagates through...). Is there something I'm missing? Would you have any idea what happened?

(DspReal tests w/o my TestModule seem to still pass -- just sanity checking some of your old specs...).

**Comment by chick** [comment]
@shunshou I'm looking at it, no clue yet

**Comment by shunshou** [comment]
It's likely a bug in my TestModule, but I've been staring at this forever and not getting very far... On the upside, it looks like Matrix ops for synthesizable stuff check out, so I'll run stuff through your instrumenting stuff (or you can too) later tomorrow and try pushing through synthesis.

**Comment by chick** [comment]
Not sure if this helps but here is the interpreter calculation of an output
```
peek io_out_0_node
...
resolve dependencies
  evaluate     io_out_0_node <= dut.io_out_0_node
      evaluate     dut.io_out_0_node <= dut.add_depth1_20_elB_0_0_node
          evaluate     dut.add_depth1_20_elB_0_0_node <= dut.add_depth1_4_elB_0_0_node  @[MatMul.scala 90:22]
              evaluate     dut.add_depth1_4_elB_0_0_node <= dut.add_depth1_elB_0_0_node  @[MatMul.scala 90:22]
                  evaluate     dut.add_depth1_elB_0_0_node <= dut._T_1751_node  @[MatMul.scala 90:22]
                      evaluate     dut._T_1751_node <= dut.BBFAdd.out  @[DspReal.scala 28:19]
                          evaluate     dut.BBFAdd.out <= BlackBoxOutput(out,UIntType(IntWidth(64)))
                            got a black box, BlackBoxOutput(out,dsptools.numbers.DspRealAdd@34ff3c9f,List(dut.BBFAdd.in1, dut.BBFAdd.in2),UIntType(IntWidth(64)))
                              evaluate     dut.BBFAdd.in1 <= dut.split11_depth0_4_elB_0_0_node
                                  evaluate     dut.split11_depth0_4_elB_0_0_node <= dut.split11_depth0_2_elB_0_0_node  @[MatMul.scala 90:22]
                                      evaluate     dut.split11_depth0_2_elB_0_0_node <= dut.split11_depth0_elB_0_0_node  @[MatMul.scala 90:22]
                                          evaluate     dut.split11_depth0_elB_0_0_node <= dut.a_elB_0_0_node  @[MatMul.scala 90:22]
                                              evaluate     dut.a_elB_0_0_node <= dut.io_a_0_node  @[MatMul.scala 34:22]
                                                  evaluate     dut.io_a_0_node <= io_a_0_node
                                                      evaluate     io_a_0_node <= UInt<64>("h0")
                                                      evaluated    io_a_0_node <= 0.U<64>
                                                  evaluated    dut.io_a_0_node <= 0.U<64>
                                              evaluated    dut.a_elB_0_0_node <= 0.U<64>
                                          evaluated    dut.split11_depth0_elB_0_0_node <= 0.U<64>
                                      evaluated    dut.split11_depth0_2_elB_0_0_node <= 0.U<64>
                                  evaluated    dut.split11_depth0_4_elB_0_0_node <= 0.U<64>
                              evaluated    dut.BBFAdd.in1 <= 0.U<64>
                              evaluate     dut.BBFAdd.in2 <= dut.split11_depth0_5_elB_0_0_node
                                  evaluate     dut.split11_depth0_5_elB_0_0_node <= dut.split11_depth0_3_elB_0_0_node  @[MatMul.scala 90:22]
                                      evaluate     dut.split11_depth0_3_elB_0_0_node <= dut.split11_depth0_1_elB_0_0_node  @[MatMul.scala 90:22]
                                          evaluate     dut.split11_depth0_1_elB_0_0_node <= dut.b_elB_0_0_node  @[MatMul.scala 90:22]
                                              evaluate     dut.b_elB_0_0_node <= dut.io_b_0_node  @[MatMul.scala 34:22]
                                                  evaluate     dut.io_b_0_node <= io_b_0_node
                                                      evaluate     io_b_0_node <= UInt<64>("h0")
                                                      evaluated    io_b_0_node <= 0.U<64>
                                                  evaluated    dut.io_b_0_node <= 0.U<64>
                                              evaluated    dut.b_elB_0_0_node <= 0.U<64>
                                          evaluated    dut.split11_depth0_1_elB_0_0_node <= 0.U<64>
                                      evaluated    dut.split11_depth0_3_elB_0_0_node <= 0.U<64>
                                  evaluated    dut.split11_depth0_5_elB_0_0_node <= 0.U<64>
                              evaluated    dut.BBFAdd.in2 <= 0.U<64>
                          evaluated    dut.BBFAdd.out <= 0.U<64>
                      evaluated    dut._T_1751_node <= 0.U<64>
                  evaluated    dut.add_depth1_elB_0_0_node <= 0.U<64>
              evaluated    dut.add_depth1_4_elB_0_0_node <= 0.U<64>
          evaluated    dut.add_depth1_20_elB_0_0_node <= 0.U<64>
      evaluated    dut.io_out_0_node <= 0.U<64>
  evaluated    io_out_0_node <= 0.U<64>
peek io_out_0_node  0
firrtl>>
```

**Comment by shunshou** [comment]
@chick @azidar @grebe i'm gonna be making some changes to my FFT code that depends on this branch... Intervals + everything associated with them should get merged at some point. Any chance you guys can take a look to see what is needed?

**Comment by chick** [comment]
I'm back from vacation and will look into this this week.

**Comment by grebe** [comment]
I can take a look at things.

**Comment by shunshou** [comment]
@chick Can you look into two things for me? I'm on a really dated branch of Chisel so I don't know if these have been resolved:

1) Can you Mux between two subclasses of Bundle? I wrote a MixedRadix class that extends Bundle, but when I try Mux(cond, mixedRadA, mixedRadB), I get an error:

```
java.lang.IllegalArgumentException: requirement failed: can't create Mux with non-equivalent types dsptools.numbers.representations.MixedRadix@104 and dsptools.numbers.representations.MixedRadix@e5
```

I think I had this issue before -- instead, I have to manually mux elements of the Bundle... 

------

2) 

```RegEnable(countNextWithWrap, init = init, enable = io.enable)```

If countNextWithWrap and init are both my MixedRadix types, it looks like the Reg output always gets the width of init. This is problematic when I set the elements of init to 0.U (width = 1)... It should take the max width of the elements of countNextWithWrap and init. 

Again, not sure if these have been addressed in newer versions of chisel...

**Comment by chick** [comment]
@shunshou 
Q: Can you Mux between two subclasses of Bundle? Ex: a MixedRadix class that extends Bundle, but when I try Mux(cond, mixedRadA, mixedRadB), I get an error:
A: It looks to me like this has not been changed or fixed. The requirement is based on typeEquivalent() and it does not seem to have been changed. Do you have a small example that I can test with. What should it do, and what would the output type be?

**Comment by shunshou** [comment]
https://github.com/ucb-bar/dsptools/blob/cbfb214373b33edad54a58fe18eebd8caec00177/src/main/scala/dsptools/numbers/representations/MixedRadixCounter.scala#L90

See the commented line. counterNextWithWrapInt mimics the behavior, but it works. 

To get the Reg in line 101 to work, I had to do lit.U(rad.W) on line 61; otherwise the counter wouldn't count.

**Comment by chick** [comment]
Closing this for now. Can be revived as needed

**Pull request closed without merging**